### PR TITLE
hdf5: conflicts zlib-ng+new_strategies

### PR DESCRIPTION
**Pull request by skosukhin**
The package should not be built against `zlib-ng+new_strategies` (see https://github.com/HDFGroup/hdf5/pull/4147). As far as I understand, the fix is backported and will be available in `@1.14.4`.

### CODE REVIEW

**Comment by skosukhin** [comment]
I would prefer the following instead:
```python
conflicts("^zlib-ng+new_strategies", when="@:1.14.3,develop-1.8:develop-1.12")
```
It concretizes fine for `spack spec hdf5`, i.e. it gives `hdf5@1.14.3 ^zlib-ng~new_strategies`.
However, `spack spec netcdf-c` prefers the `develop` version of `hdf5`: `netcdf-c ^hdf5@develop-1.15 ^zlib-ng+new_strategies`, which is technically correct but undesirable.

@alalazo do you think there is anything reasonable that can be done so that the `develop` versions are deprioritized (even more)?

File: var/spack/repos/builtin/packages/hdf5/package.py [hunk 1]
```diff
@@ -156,6 +156,9 @@ class Hdf5(CMakePackage):
         "+fortran", when="@1.13.3:^cmake@:3.22", msg="cmake_minimum_required is not set correctly."
     )
 
+    # See https://github.com/HDFGroup/hdf5/pull/4147
+    requires("^zlib-ng~new_strategies", when="@:1.14.3,develop-1.8:develop-1.12 ^zlib-ng")
+
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/hdf5/package.py]: Can you try:
```
depends_on("zlib-ng~new_strategies", when="@:1.14.3,develop-1.8:develop-1.12 ^[virtuals=zlib-api] zlib-ng")
```
?

Comment (skosukhin) [reply var/spack/repos/builtin/packages/hdf5/package.py]: It works that way, yes. I have to say though that it feels a bit like playing a lottery.

Comment (alalazo) [reply var/spack/repos/builtin/packages/hdf5/package.py]: :sweat: Yeah, I can understand that. Let's say that the way this is implemented right now, using `depends_on` or `requires` for this edge case gives the same set of possible solutions, but orders them differently (so the optimal one is different).

Comment (tgamblin) [reply var/spack/repos/builtin/packages/hdf5/package.py]: did you ever agree on what to do about this?  Are we going with @alalazo's suggestion or @skosukhin's?

Comment (skosukhin) [reply var/spack/repos/builtin/packages/hdf5/package.py]: I'm fine with following @alalazo's suggestion. I just need to find time to update the PR.

Comment (skosukhin) [reply var/spack/repos/builtin/packages/hdf5/package.py]: @alalazo I guess we need some extra magic here to pass the gitlab-ci. Could you help with that, please?

**Comment by alalazo** [comment]
@spackbot run pipeline

File: var/spack/repos/builtin/packages/hdf5/package.py (hunk 1 above)
Comment (alalazo) [reply var/spack/repos/builtin/packages/hdf5/package.py]: I tried re-running the pipelines. Hopefully they are just "transient" errors from the Gitlab cluster.

File: var/spack/repos/builtin/packages/hdf5/package.py [hunk 2]
```diff
@@ -123,7 +123,13 @@ class Hdf5(CMakePackage):
 
     depends_on("java", type=("build", "run"), when="+java")
     depends_on("szip", when="+szip")
+
     depends_on("zlib-api")
+    # See https://github.com/HDFGroup/hdf5/pull/4147
```
Comment (hyoklee) [inline var/spack/repos/builtin/packages/hdf5/package.py:128]: See also https://github.com/HDFGroup/hdf5/pull/4935

**Review by hyoklee — APPROVED** [review]

File: var/spack/repos/builtin/packages/hdf5/package.py (hunk 1 above)
Comment (skosukhin) [reply var/spack/repos/builtin/packages/hdf5/package.py]: I assume that due to the changes in the PR, different dependents of `zlib-ng` expect the package in different configurations, which the Windows job doesn't like.

Comment (alalazo) [reply var/spack/repos/builtin/packages/hdf5/package.py]: @skosukhin Unfortunately it's not like that.  Now it seems some spec is indexed by the Win buildcache (so marked as "already available as a binary"), but cannot be retrieved by the build job. Last week this was a transient error on the Win runners. Let me try again :cry:

Comment (alalazo) [reply var/spack/repos/builtin/packages/hdf5/package.py]: For the record, and example of the error is https://gitlab.spack.io/spack/spack/-/jobs/12900489

Screenshot from 2024-10-09 11-34-37

**Comment by alalazo** [comment]
@spackbot run pipeline

**Comment by alalazo** [comment]
@johnwparent @kwryankrattiger Fine with you to force merge this PR? The Windows failure on Gitlab is unrelated. Or is there any way to make this pass Gitlab CI?

**Comment by kwryankrattiger** [comment]
Since this is changing HDF5 and the config for zlib-ng we are using, I would like to see it build and run tests before merge since this is a core library, especially for the windows pipeline.

The issue is related to PR build cache. The fix should be to simply close and reopen the PR and let CI run.

**Comment by johnwparent** [comment]
Agreed, would also like to see any validation that this passes on Windows. 

> The issue is related to PR build cache. The fix should be to simply close and reopen the PR and let CI run.

That was not sufficient in my experience, I had to create a new PR, as closing cleared the build cache, but generate was continuing to generate the same hashes, so every stack started missing.

**Comment by kwryankrattiger** [comment]
@spackbot run pipeline

**Comment by kwryankrattiger** [comment]
You shouldn't need a new PR, as the spackbot reply shows the branch is gone on gitlab as well, so sync will make a new merge commit with the latest develop and push that. The close PR action makes the re-opened PR look the same as a new PR with the same commit. If the robot doesn't pick up the event then this wouldn't work, which may happen sometimes.

**Comment by kwryankrattiger** [comment]
Once this finishes we will need to re-run again :/ not sure why but gitlab is dropping trigger pipelines which is a new one for me.

**Comment by kwryankrattiger** [comment]
@spackbot run pipeline

**Review by kwryankrattiger — APPROVED** [review]
It looks like the windows are past the zlib-ng/HDF5 issues.

The other pipelines are also looking okay, but will probably take a while since we cleared all of the new stuff from the PR cache.

Let me know if this is still stuck, otherwise LGTM!
